### ISSUE TITLE
Support `vec![const { ... }; n]` syntax for creating a `Vec` of non-`Clone` values

### ISSUE THREAD

**Issue by EFanZh** [rust-lang/libs-team#484]
# Proposal

## Problem statement

For array types, I can use `[const { ... }; N]` syntax to create an array of `const { ... }`, even if it’s does not implement `Clone`. But I can’t do the same thing with `Vec`: `vec![const { ... }; n]` currently does not compile.

## Motivating examples or use cases

Imagining implementing a DFS algorithm with an array to save the traversal states, I can write something like:

```rust
enum State {
    NotVisited,
    Visiting,
    Visited,
}

let states = std::iter::repeat_with(|| State::NotVisited).take(n).collect::<Vec<_>>();
```

But it could be more simple if I could just write:

```rust
let states = vec![const { State::NotVisited }; n];
```

Deriving `Clone` might not always be possible because the value type could be from a third party crate.

## Solution sketch

Reimplement `vec!` macro so it supports the `const { ... }` syntax above.

## What happens now?

This issue contains an API change proposal (or ACP) and is part of the libs-api team [feature lifecycle]. Once this issue is filed, the libs-api team will review open proposals as capability becomes available. Current response times do not have a clear estimate, but may be up to several months.

[feature lifecycle]: https://std-dev-guide.rust-lang.org/development/feature-lifecycle.html

## Possible responses

The libs team may respond in various different ways. First, the team will consider the *problem* (this doesn't require any concrete solution or alternatives to have been proposed):

- We think this problem seems worth solving, and the standard library might be the right place to solve it.
- We think that this probably doesn't belong in the standard library.

Second, if there's a concrete solution:

- We think this specific solution looks roughly right, approved, you or someone else should implement this. (Further review will still happen on the subsequent implementation PR.)
- We're not sure this is the right solution, and the alternatives or other materials don't give us enough information to be sure about that. Here are some questions we have that aren't answered, or rough ideas about alternatives we'd want to see discussed.

**Comment by dtolnay** [comment]
Seems like a good idea to me. I expect the implementation you have in mind involves expanding to `<[_]>::into_vec(Box::new([const { expr }; N]))` in the const case. I think that's fine to do but it won't benefit from the good suggestion that array syntax provides where it tells you to wrap your expression in an inline const block. Compare:

```console
error[E0277]: the trait bound `State: Copy` is not satisfied
 --> src/main.rs:9:14
  |
9 |     let _ = [State::NotVisited; N];
  |              ^^^^^^^^^^^^^^^^^ the trait `Copy` is not implemented for `State`
  |
  = note: the `Copy` trait is required because this value will be copied for each element of the array
help: consider annotating `State` with `#[derive(Copy)]`
  |
2 + #[derive(Copy)]
3 | enum State {
  |
help: create an inline `const` block
  |
9 |     let _ = [const { State::NotVisited }; N];
  |              ~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

vs:

```console
error[E0277]: the trait bound `State: Clone` is not satisfied
    --> src/main.rs:9:18
     |
9    |     let _ = vec![State::NotVisited; N];
     |             -----^^^^^^^^^^^^^^^^^----
     |             |    |
     |             |    the trait `Clone` is not implemented for `State`
     |             required by a bound introduced by this call
     |
note: required by a bound in `from_elem`
    --> $RUSTUP_HOME/toolchains/nightly-aarch64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/vec/mod.rs:3172:21
     |
3172 | pub fn from_elem<T: Clone>(elem: T, n: usize) -> Vec<T> {
     |                     ^^^^^ required by this bound in `from_elem`
help: consider annotating `State` with `#[derive(Clone)]`
     |
2    + #[derive(Clone)]
3    | enum State {
     |
```

So while I don't think the implementation of this needs to be blocked on a better diagnostic, I would encourage you to think about what it would take to hint for the user to write `vec![const { expr }; N]` when they wrote `vec![expr; N]` on a const non-Clone expr.

**Comment by EFanZh** [comment]
I don’t think the array syntax can be used to implement the `vec!` macro, since the array syntax requires the length being a const value, while the length used in `vec!` can be a runtime value.

**Comment by cuviper** [comment]
Other examples where `Clone` may not be wanted or intended:
* `vec![Vec::with_capacity(cap); n]`
* `vec![Arc::new(Mutex::new(x)); n]`

**Comment by dtolnay** [comment]
> I don’t think the array syntax can be used to implement the `vec!` macro, since the array syntax requires the length being a const value

Good call; I didn't consider non-const `N` &mdash; so this would only be serving as shorthand for `iter::repeat_with(|| expr).take(N).collect::<Vec<_>>()`. I agree with Scott that the `const` doesn't feel necessary or relevant in that case.

**Comment by cuviper** [comment]
We should also be careful about how this interacts with `expr`/`expr_2021` changes:
https://doc.rust-lang.org/nightly/edition-guide/rust-2024/macro-fragment-specifiers.html

... but maybe that just means we would have to manually match the `const` before the more general `expr`.

**Comment by joshtriplett** [comment]
We talked about this in today's @rust-lang/libs-api meeting.

We were in favor of `vec![...]` to define a vector always accepting any syntax that can be used for `[...]` to define an array.

We also wondered whether this would Just Work when the standard library is converted to the 2024 edition and uses the 2024 `expr` in all macros.

**Comment by fmease** [comment]
This means `vec![const { ... }; n]` and `vec![(const { … }; n]` will differ semantically (in all editions).

**Comment by EFanZh** [comment]
> This means `vec![const { ... }; n]` and `vec![(const { … }); n]` will differ semantically (in all editions).

`vec![(const { … }); n]` calls `Clone::clone`, if we choose to make those two behave the same, `vec![const { ... }; n]` will also have call `Clone::clone`. But not all const values implement `Clone`, if we want to support non-`Clone` types, `vec![const { ... }; n]` needs to behave differently depending on whether the value implements `Clone`. Personally, I consider one syntax with two different behaviors being worse than the behavior difference caused by adding a pair of parentheses. Also, if we choose to call `Clone::clone` function with `vec![const { ... }; n]`, we lose the ability to create a `Vec` without calling the `Clone::clone` function.

Another solution would be adding a new macro that is dedicated to handing the const value case.

**Comment by kennytm** [comment]
@EFanZh It is a known design that `x` and `{x}` behaves differently (esp. regarding borrow-check), but `x` and `(x)` always behaved the same until your proposed syntax.

**Comment by traviscross** [comment]
It is also the case that I'd expect `mac!(const { $expr })` and `const X: _ = $expr; mac!(X)` to have the same behavior.

**Comment by programmerjake** [comment]
if we want `[MY_NAMED_CONST; N]` syntax to work with things other than arrays, maybe we need a built-in macro `is_expr_copyable!` that is true if the passed-in expression is a `const` block/item or is a type that implements `Copy`. (this isn't specialization, it only works on types where `[v; 2]` would work for an expression `v` with that type, so this generally requires a visible `T: Copy` bound.)

**Comment by cuviper** [comment]
I think it would be better to have an explicit syntax for "repeat this expression" (regardless of `const`) vs. the current cloning, whether that's another variation of `vec!` or a new macro or method -- assuming we can do better than `iter::repeat_with`.

**Comment by EFanZh** [comment]
`Vec::from_fn` is implemented with tracking issue rust-lang/rust#149698.
